File: examples/rust/iris.rs
#![no_std]
    pub mod packages {
        pub mod Telemetry {
            #[derive(Copy, Clone)]
            pub enum Status {
                SLEEP = 0,
                ACTIVE = 1,
                FLIGHT = 2,
                IDLE = 3,
            }
            impl Status {
                pub fn to_be_bytes(&self) -> [u8; 4] {
                    (match self {
                        Status::SLEEP => 0,
                        Status::ACTIVE => 1,
                        Status::FLIGHT => 2,
                        Status::IDLE => 3,
                    } as u32)
                        .to_be_bytes()
                }
                pub fn from_be_bytes(data: [u8; 4]) -> Status {
                    match u32::from_be_bytes(data) {
                        0 => Status::SLEEP,
                        1 => Status::ACTIVE,
                        2 => Status::FLIGHT,
                        3 => Status::IDLE,
                        _ => panic!("No variant found."),
                    }
                }
            }
            #[derive(Copy, Clone)]
            pub struct Computer {
                pub id: u32,
                pub batteries: [Battery; 2],
                pub status: Status,
            }
            impl Computer {
                pub const NAME_HASH: u32 = 3613607352;
                pub const BYTES_LENGTH: usize = 16 + 4;
                pub fn encode(&self) -> [u8; Computer::BYTES_LENGTH] {
                    let mut data: [u8; Computer::BYTES_LENGTH] = [0; Computer::BYTES_LENGTH];
                    let mut index = 0;
                    for x in u32::to_be_bytes(Computer::NAME_HASH) {
                        data[index] = x;
                        index += 1;
                    }
                    for x in self.to_be_bytes() {
                        data[index] = x;
                        index += 1;
                    }
                    data
                }
                pub fn to_be_bytes(&self) -> [u8; Computer::BYTES_LENGTH - 4] {
                    let mut data: [u8; Computer::BYTES_LENGTH - 4] =
                        [0; Computer::BYTES_LENGTH - 4];
                    let mut index = 0;
                    for x in self.id.to_be_bytes() {
                        data[index] = x;
                        index += 1;
                    }
                    for i in self.batteries {
                        for x in i.to_be_bytes() {
                            data[index] = x;
                            index += 1;
                        }
                    }
                    for x in self.status.to_be_bytes() {
                        data[index] = x;
                        index += 1;
                    }
                    data
                }
                pub fn decode(data: &[u8]) -> Computer {
                    Computer::from_be_bytes(data[4..data.len()].try_into().unwrap())
                }
                pub fn from_be_bytes(data: [u8; Computer::BYTES_LENGTH - 4]) -> Computer {
                    let mut out = Computer {
                        id: 0,
                        batteries: [Battery { charge: 0.0 }; 2],
                        status: Status::SLEEP,
                    };
                    let mut index = 0;
                    out.id = u32::from_be_bytes(data[index..index + 4].try_into().unwrap());
                    index += 4;
                    for i in 0..2 {
                        out.batteries[i] =
                            Battery::from_be_bytes(data[index..index + 4].try_into().unwrap());
                        index += 4;
                    }
                    out.status = Status::from_be_bytes(data[index..index + 4].try_into().unwrap());
                    index += 4;
                    out
                }
            }
            #[derive(Copy, Clone)]
            pub struct Data {
                pub computers: [Computer; 2],
            }
            impl Data {
                pub const NAME_HASH: u32 = 1062369733;
                pub const BYTES_LENGTH: usize = 32 + 4;
                pub fn encode(&self) -> [u8; Data::BYTES_LENGTH] {
                    let mut data: [u8; Data::BYTES_LENGTH] = [0; Data::BYTES_LENGTH];
                    let mut index = 0;
                    for x in u32::to_be_bytes(Data::NAME_HASH) {
                        data[index] = x;
                        index += 1;
                    }
                    for x in self.to_be_bytes() {
                        data[index] = x;
                        index += 1;
                    }
                    data
                }
                pub fn to_be_bytes(&self) -> [u8; Data::BYTES_LENGTH - 4] {
                    let mut data: [u8; Data::BYTES_LENGTH - 4] = [0; Data::BYTES_LENGTH - 4];
                    let mut index = 0;
                    for i in self.computers {
                        for x in i.to_be_bytes() {
                            data[index] = x;
                            index += 1;
                        }
                    }
                    data
                }
                pub fn decode(data: &[u8]) -> Data {
                    Data::from_be_bytes(data[4..data.len()].try_into().unwrap())
                }
                pub fn from_be_bytes(data: [u8; Data::BYTES_LENGTH - 4]) -> Data {
                    let mut out = Data {
                        computers: [Computer {
                            status: Status::SLEEP,
                            batteries: [Battery { charge: 0.0 }; 2],
                            id: 0,
                        }; 2],
                    };
                    let mut index = 0;
                    for i in 0..2 {
                        out.computers[i] =
                            Computer::from_be_bytes(data[index..index + 16].try_into().unwrap());
                        index += 16;
                    }
                    out
                }
            }
            #[derive(Copy, Clone)]
            pub struct Battery {
                pub charge: f32,
            }
            impl Battery {
                pub const NAME_HASH: u32 = 2215305518;
                pub const BYTES_LENGTH: usize = 4 + 4;
                pub fn encode(&self) -> [u8; Battery::BYTES_LENGTH] {
                    let mut data: [u8; Battery::BYTES_LENGTH] = [0; Battery::BYTES_LENGTH];
                    let mut index = 0;
                    for x in u32::to_be_bytes(Battery::NAME_HASH) {
                        data[index] = x;
                        index += 1;
                    }
                    for x in self.to_be_bytes() {
                        data[index] = x;
                        index += 1;
                    }
                    data
                }
                pub fn to_be_bytes(&self) -> [u8; Battery::BYTES_LENGTH - 4] {
                    let mut data: [u8; Battery::BYTES_LENGTH - 4] = [0; Battery::BYTES_LENGTH - 4];
                    let mut index = 0;
                    for x in self.charge.to_be_bytes() {
                        data[index] = x;
                        index += 1;
                    }
                    data
                }
                pub fn decode(data: &[u8]) -> Battery {
                    Battery::from_be_bytes(data[4..data.len()].try_into().unwrap())
                }
                pub fn from_be_bytes(data: [u8; Battery::BYTES_LENGTH - 4]) -> Battery {
                    let mut out = Battery { charge: 0.0 };
                    let mut index = 0;
                    out.charge = f32::from_be_bytes(data[index..index + 4].try_into().unwrap());
                    index += 4;
                    out
                }
            }
        }
    }
    pub enum Structs {
        Telemetry_Computer(packages::Telemetry::Computer),
        Telemetry_Data(packages::Telemetry::Data),
        Telemetry_Battery(packages::Telemetry::Battery),
    }
    pub fn decode(data: &[u8]) -> Result<Structs, &str> {
        let struct_name_hash = u32::from_be_bytes(data[0..4].try_into().unwrap());
        match struct_name_hash {
            packages::Telemetry::Computer::NAME_HASH
                if data.len() == packages::Telemetry::Computer::BYTES_LENGTH =>
            {
                Ok(Structs::Telemetry_Computer(
                    packages::Telemetry::Computer::decode(&data),
                ))
            }
            packages::Telemetry::Data::NAME_HASH
                if data.len() == packages::Telemetry::Data::BYTES_LENGTH =>
            {
                Ok(Structs::Telemetry_Data(packages::Telemetry::Data::decode(
                    &data,
                )))
            }
            packages::Telemetry::Battery::NAME_HASH
                if data.len() == packages::Telemetry::Battery::BYTES_LENGTH =>
            {
                Ok(Structs::Telemetry_Battery(
                    packages::Telemetry::Battery::decode(&data),
                ))
            }
            _ => Err("Unknown data."),
        }
    }

File: examples/python/iris.py
import enum
import struct
import itertools



class Iris:
    class Packages:
        class Telemetry:
            class Status(enum.IntEnum):
                SLEEP = 0
                ACTIVE = 1
                FLIGHT = 2
                IDLE = 3
                def to_be_bytes(self) -> bytes:
                    return struct.pack('>I', self)
                @staticmethod
                def from_be_bytes(raw: bytes):
                    match struct.unpack('>I', raw)[0]:
                        case 0:
                            return Iris.Packages.Telemetry.Status.SLEEP
                        case 1:
                            return Iris.Packages.Telemetry.Status.ACTIVE
                        case 2:
                            return Iris.Packages.Telemetry.Status.FLIGHT
                        case 3:
                            return Iris.Packages.Telemetry.Status.IDLE
                        case _:
                            raise 'No variant found.'
            class Computer:
                NAME_HASH = 3613607352
                BYTES_LENGTH = 16 + 4
                def __init__(self, id, batteries, status):
                    self.id = id
                    self.batteries = batteries
                    self.status = status
                def encode(self) -> bytes:
                    return struct.pack('>I16B', self.NAME_HASH, *self.to_be_bytes())
                def to_be_bytes(self) -> bytes:
                    return struct.pack('>I8BI', self.id, *itertools.chain.from_iterable([i.to_be_bytes() for i in self.batteries]), self.status)
                @staticmethod
                def decode(raw: bytes):
                    data = struct.unpack('>I16B', raw)
                    return Iris.Packages.Telemetry.Computer.from_be_bytes(bytes(data[1:]))
                @staticmethod
                def from_be_bytes(raw: bytes):
                    data = struct.unpack('>I8BI', raw)
                    return Iris.Packages.Telemetry.Computer(
                        id=data[0],
                        batteries=[Iris.Packages.Telemetry.Battery.from_be_bytes(bytes(data[i:i+4])) for i in range(1, 8, 4)],
                        status=data[9]
                    )
            class Data:
                NAME_HASH = 1062369733
                BYTES_LENGTH = 32 + 4
                def __init__(self, computers) -> None:
                    self.computers = computers
                def encode(self) -> bytes:
                    return struct.pack('>I32B', self.NAME_HASH, *self.to_be_bytes())
                def to_be_bytes(self) -> bytes:
                    return struct.pack('>32B', *itertools.chain.from_iterable([i.to_be_bytes() for i in self.computers]))
                @staticmethod
                def decode(raw: bytes):
                    data = struct.unpack('>I32B', raw)
                    return Iris.Packages.Telemetry.Data.from_be_bytes(bytes(data[1:]))
                @staticmethod
                def from_be_bytes(raw: bytes):
                    data = struct.unpack('>32B', raw)
                    return Iris.Packages.Telemetry.Data(
                        computers=[Iris.Packages.Telemetry.Computer.from_be_bytes(bytes(data[i:i+16])) for i in range(0, 32, 16)]
                    )
            class Battery:
                NAME_HASH = 2215305518
                BYTES_LENGTH = 4 + 4
                def __init__(self, charge) -> None:
                    self.charge = charge
                def encode(self) -> bytes:
                    return struct.pack('>I4B', self.NAME_HASH, *self.to_be_bytes())
                def to_be_bytes(self) -> bytes:
                    return struct.pack('>f', self.charge)
                @staticmethod
                def decode(raw: bytes):
                    data = struct.unpack('>I4B', raw)
                    return Iris.Packages.Telemetry.Battery.from_be_bytes(bytes(data[1:]))
                @staticmethod
                def from_be_bytes(raw: bytes):
                    data = struct.unpack('>f', raw)
                    return Iris.Packages.Telemetry.Battery(
                        charge=data[0]
                    )
    @staticmethod
    def decode(raw: bytes):
        name_hash = struct.unpack('>I', raw[0:4])[0]
        match name_hash:
            case Iris.Packages.Telemetry.Computer.NAME_HASH if len(raw) == Iris.Packages.Telemetry.Computer.BYTES_LENGTH:    
                return Iris.Packages.Telemetry.Computer.decode(raw)
            case Iris.Packages.Telemetry.Data.NAME_HASH if len(raw) == Iris.Packages.Telemetry.Data.BYTES_LENGTH:
                return Iris.Packages.Telemetry.Data.decode(raw)
            case Iris.Packages.Telemetry.Battery.NAME_HASH if len(raw) == Iris.Packages.Telemetry.Battery.BYTES_LENGTH:
                return Iris.Packages.Telemetry.Battery.decode(raw)
            case _:
                raise 'Unknown data.'
            

d = Iris.Packages.Telemetry.Data(
    computers=[
        Iris.Packages.Telemetry.Computer(
            id=123,
            batteries=[
                Iris.Packages.Telemetry.Battery(
                    charge=43.65
                ),
                Iris.Packages.Telemetry.Battery(
                    charge=88.32
                )
            ],
            status=Iris.Packages.Telemetry.Status.IDLE
        ),
        Iris.Packages.Telemetry.Computer(
            id=1337,
            batteries=[
                Iris.Packages.Telemetry.Battery(
                    charge=32.88
                ),
                Iris.Packages.Telemetry.Battery(
                    charge=123.312
                )
            ],
            status=Iris.Packages.Telemetry.Status.FLIGHT
        )
    ]
)

en = d.encode()
print([int(i) for i in en])

de = Iris.decode(en)

if isinstance(de, Iris.Packages.Telemetry.Data):
    print('TELEMETRY DATA')
    for c in de.computers:
        print('COMPUTER', c.id, c.status)
        print('BATTERIES:')
        for b in c.batteries:
            print(b.charge)
elif isinstance(de, Iris.Packages.Telemetry.Computer):
    pass
elif isinstance(de, Iris.Packages.Telemetry.Battery):
    pass
else:
    raise 'Error while decoding.'